### PR TITLE
 Software/Devices Macro #26 

### PR DESCRIPTION
**Pull request by Farcasut**
Added 8 new macros:
-BrowserEngines:
image
-Browsers
image
-Configurations
image
-DeviceBrand:
image
-DeviceModel:
image
-DeviceType:
image
-OperatingSystemVersions:
image
-ScreenResolution
image

### CODE REVIEW

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Configurations.xml [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+  'meta': {
+    'propertyDescriptors': [
+      {'id': 'configurations', 'displayer': 'html'},
+      {'id': 'visits', 'displayer': 'html'},
+      {'id': 'nbOfActions', 'displayer': 'html'},
+      {'id': 'timeSpent', 'displayer': 'html'},
+      {'id': 'bounceRate', 'displayer': 'html'},
+      {'id': 'conversionRate', 'displayer': 'html'},
+      {'id': 'actions',"filterable": false, "sortable": false, 'displayer': 'html'}
+    ],
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Configurations.xml:268]: Single quotes
Below as well

**Review by oanalavinia — CHANGES_REQUESTED** [review]
There are some codestyle changes needed for all files, I only added the comment for some of them

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJSON.xml [hunk 2]
```diff
@@ -41,13 +41,39 @@
   'period' : $request.period,
   'date' : $request.date,
   'module' : 'API',
-  'method' : 'Actions.getPageUrl',
   'format' : 'json'
 })
+#set ($filters = {})
 #if ($request.macroName == 'MostViewedPages')
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJSON.xml:47]: What do you say about refactoring to something like this:
```
#set ($macrosParams = {
  'MostViewedPages': {
    'method': 'Actions.getPageUrl',
    'rowIdentifierName': 'pageUrl',
    'rowIdentifierValue': $request.rowIdentifier
  }, 'Browser': {
    'method': 'DevicesDetection.getModel',
    'rowIdentifierName': 'segment',
    'rowIdentifierValue': $request.rowIdentifier
  },
  ...
})
#set ($parameters = {})
#set ($macroParams = $macrosParams.get($request.macroName))
#set ($discard = $parameters.put($macroParams.rowIdentifierName, $macroParams.rowIdentifierValue))
#set ($discard = $parameters.put('method', $macroParams.method))
```

Comment (oanalavinia) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJSON.xml:47]: Actually, I noticed now on [PageFollowingASiteSearch](https://github.com/xwikisas/application-analytics/pull/30/files#diff-a03c3eeff3a8ec55bd955ee5dbe8313b607bae39af96eca19f155d8b59aaa385R50) that there can be multiple parameters to add, so this is not actually helpfull

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJSON.xml [hunk 3]
```diff
@@ -41,13 +41,39 @@
   'period' : $request.period,
   'date' : $request.date,
   'module' : 'API',
-  'method' : 'Actions.getPageUrl',
   'format' : 'json'
 })
+#set ($filters = {})
```
[the remark this answers is not in the document]
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJSON.xml:46]: In most cases yes, but when someone is using the pages following a site search macro the filters would be populated with the page url

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Configurations.xml [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+{{liveData
+  id="$macroId"
+  properties='configurations, visits, nbOfActions, timeSpent, bounceRate, conversionRate, actions'
+  source='liveTable'
+  sourceParameters="$parameters"
+  sort='visits:desc'
+}}$jsontool.serialize($liveDataConfig){{/liveData}}
+
+#set ($rowEvolutionParams = {
+  $'visits': 'nb_visits',
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Configurations.xml:283]: What's with that `$`?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    'returnedrows': $resultList.size(),
+    'offset': $mathtool.add($offset, 1),
+    'reqNo': $numbertool.toNumber($request.reqNo).intValue(),
+    'rows': []
+  })
+  #foreach ($currentEntry in $resultList)
+    ## The Resolution.getConfiguration method from matomo doesn't return a segment for each row and to get segment I
+    ## must split the label into it's components and construct the segment myself.
+    #if ($request.matomoMethod == 'Resolution.getConfiguration')
+    #set ($values= $currentEntry.get('label').asText().split(' / '))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml:107]: Indentation (for the whole block)

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    #set ($segmentsName = ['operatingSystemName', 'browserName', 'resolution'])
+    #set ($finalSegment = '')
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml:109]: You could have a list where to add values and do a `$stringtool.join($segmentParts, ';')` at the end

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    #foreach ($i in [0..2])
+      #set ($currentSegment = $segmentsName.get($i) + '==' + $values.get($i))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml:111]: So you know that the received values are in order for `'operatingSystemName', 'browserName', 'resolution'`?

**Review by oanalavinia — CHANGES_REQUESTED** [review]
You should also add a licensing check on each macro page.
Otherwise, I added just some questions:)

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 8]
```diff
@@ -74,6 +74,28 @@ analytics.rowEvolution.modal.year.label=Year
 analytics.rowEvolution.modal.field.pageViews=Page views
 analytics.rowEvolution.modal.field.uniqueViews=Unique views
 analytics.rowEvolution.modal.field.avgTime=Avg. Time spent
+analytics.rowEvolution.modal.field.clickedSearchResult=Clicked in search results
+analytics.rowEvolution.modal.field.visits=Visits
+analytics.rowEvolution.modal.field.actions=Actions
+analytics.rowEvolution.modal.field.timeSpent=Time spent
+analytics.rowEvolution.modal.field.bounceRate=Bounce Rate
+analytics.rowEvolution.modal.field.conversionRate=Conversion Rate
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:82]: Rate should be in lowercase

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 9]
```diff
@@ -74,6 +74,28 @@ analytics.rowEvolution.modal.year.label=Year
 analytics.rowEvolution.modal.field.pageViews=Page views
 analytics.rowEvolution.modal.field.uniqueViews=Unique views
 analytics.rowEvolution.modal.field.avgTime=Avg. Time spent
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:76]: Time shpuld be in lowercase

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml (hunk 7 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml:111]: Yes, the order will always be the same

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      #set ($values= $currentEntry.get('label').asText().split(' / '))
+      #set ($segmentsName = ['operatingSystemName', 'browserName', 'resolution'])
+      #set ($segmentParts = [])
+      #foreach ($i in [0..2])
+        #set ($currentSegment = $segmentsName.get($i) + '==' + $values.get($i))
+        #set ($discard = $segmentParts.add($currentSegment))
+      #end
+      #set ($segment = $stringtool.join($segmentParts, ';'))
+      #else
+        #set ($segment = $currentEntry.get('segment').asText())
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/SoftwareDevicesJSON.xml:116]: Indentation (2 spaces too much)

**Review by oanalavinia — APPROVED** [review]
Just one comment, you can merge it after

**Review by oanalavinia — CHANGES_REQUESTED** [review]
Missing licensing check for macros

**Review by oanalavinia — APPROVED** [review]

**Pull request merged**